crates/re_log_types/src/data_cell.rs

```diff
@@ -0,0 +1,445 @@
+    UnsupportedDatatype(arrow2::datatypes::DataType),
+
+    #[error("Could not serialize/deserialize component instances to/from Arrow")]
+    Arrow(#[from] arrow2::error::Error),
+}
+
+pub type DataCellResult<T> = ::std::result::Result<T, DataCellError>;
+
+// ---
+
+/// A cell's worth of data, i.e. a uniform array of values for a given component type.
+/// This is the leaf type in our data model.
+///
+/// A `DataCell` can be constructed from either an iterable of native `Component`s or directly
+/// from a slice of arrow data.
+///
+/// Behind the scenes, a `DataCell` is backed by an erased arrow array living on the heap, which
+/// is likely to point into a larger batch of contiguous memory that it shares with its peers.
+/// Cloning a `DataCell` is thus cheap (shallow, ref-counted).
+///
+/// ## Layout
+///
+/// A cell is an array of component instances: `[C, C, C, ...]`.
+///
+/// Consider this example:
+/// ```ignore
+/// let points: &[Point2D] = &[
+///     [10.0, 10.0].into(),
+///     [20.0, 20.0].into(),
+///     [30.0, 30.0].into(),
+/// ];
+/// let cell = DataCell::from(points);
+/// // Or, alternatively:
+/// let cell = DataCell::from_component::<Point2D>([[10.0, 10.0], [20.0, 20.0], [30.0, 30.0]]);
+/// ```
+///
+/// The cell's datatype is now a `StructArray`:
+/// ```ignore
+/// Struct([
+///    Field { name: "x", data_type: Float32, is_nullable: false, metadata: {} },
+///    Field { name: "y", data_type: Float32, is_nullable: false, metadata: {} },
+/// ])
+/// ```
+///
+/// Or, visualized as a cell within a larger table:
+/// ```text
+/// ┌──────────────────────────────────────────────────┐
+/// │ rerun.point2d                                    │
+/// ╞══════════════════════════════════════════════════╡
+/// │ [{x: 10, y: 10}, {x: 20, y: 20}, {x: 30, y: 30}] │
+/// └──────────────────────────────────────────────────┘
+/// ```
+///
+/// ## Example
+///
+/// ```rust
+/// # use arrow2_convert::field::ArrowField as _;
+/// # use itertools::Itertools as _;
+/// #
+/// # use re_log_types::{DataCell, msg_bundle::Component as _};
+/// # use re_log_types::component_types::Point2D;
+/// #
+/// let points: &[Point2D] = &[
+///     [10.0, 10.0].into(),
+///     [20.0, 20.0].into(),
+///     [30.0, 30.0].into(),
+/// ];
+/// let _cell = DataCell::from(points);
+///
+/// // Or, alternatively:
+/// let cell = DataCell::from_component::<Point2D>([[10.0, 10.0], [20.0, 20.0], [30.0, 30.0]]);
+///
+/// eprintln!("{:#?}", cell.datatype());
+/// eprintln!("{cell}");
+/// #
+/// # assert_eq!(Point2D::name(), cell.component());
+/// # assert_eq!(3, cell.num_instances());
+/// # assert_eq!(cell.datatype(), &Point2D::data_type());
+/// #
+/// # assert_eq!(points, cell.as_native().collect_vec().as_slice());
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct DataCell {
+    /// Name of the component type used in this cell.
+    //
+    // TODO(cmc): We need to store this extra information within the values array itself,
```

Comment:
Is "need" really the right word here? Why do we need it?

Saying that we need to do A and then do B is quite confusing :)